mesonbuild/dependencies/__init__.py

```diff
@@ -32,6 +32,173 @@
+        if lib is None:
+            mlog.debug('Could not find lib.')
+            self.is_found = False
+            return
+
+        self.compile_args.append(f'-I{os.path.join(root, "include")}')
+        self.link_args.append(lib)
+        self.is_found = True
+```
+
+This is nicer in a couple of ways. First we can properly cross compile as we
+are allowed to set FOO_ROOT for both the build and host machines, it also
+means that users can override this in their machine files, and finally if
+that environment variables changes durring a meson reconfigure meson won't
+re-read it.
+
+We can still make this better by handling the `static` keyword
+```python
+class FooSystemDependency(ExternalDependency):
+
+    def __init__(self, name: str, environment: 'Environment', kwargs: T.Dict[str, T.Any]):
+        super().__init__(name, environment, kwargs)
+        root = environment.properties[self.for_machine].foo_root
+        if root is None:
+            mlog.debug('foo_root is unset.')
+            self.is_found = False
+            return
+
+        static = kwargs.get('static', False)
+        static = mesonlib.LibType.STATIC if kwargs.get('static', False) else mesonlib.LibType.SHARED
+        lib = self.clib_compiler.find_library(
+            'foo', environment, [os.path.join(root, 'lib')], libtype=static)
+        if lib is None:
+            mlog.debug('Could not find lib.')
+            self.is_found = False
+            return
+
+        self.compile_args.append(f'-I{os.path.join(root, "include")}')
+        self.link_args.append(lib)
+        self.is_found = True
+```
+
+Now we meson will figure out wither it should be finding `libfoo.so` or
+`libfoo.a` (or the platform specific names). Things are looking pretty good
+now, so it can be added to the `packages` dict below
+```python
+packages.udate({
+    'foo': FooSystemDependency,
+})
+```
+
+Now, what if foo also provides pkg-config, but it's only shipped on Unices,
+or only inlucded in very recent versions? We can use the `DependencyFactory`
+class:
+```python
+foo_factory = DependencyFactory(
+    'foo',
+    [DependencyMethods.PKGCONFIG, DependencyMethods.SYSTEM],
+    system_class=FooSystemDependency,
+)
+```
+This is a helper function that will generate a default pkg-config based
+dependency, and use the `FooSystemDependency` as well. You would then add the
+`foo_factory` to packages instead
+```python
+packages.udate({
+    'foo': foo_factory,
+})
+```
+
+If you have a dependency that is very complicated, (such as having multiple
+implementations) you may need to write your own factory function. There are a
+number of examples in this pacakge.
+
+_Note_ before we moved to factory functions it was common to use an
+`ExternalDependency` class that would instantiate different types of
+dependencies and hold the one it found. There are a number of drawbacks to
+this approach, and no new dependencies should do this.
```

Comment:
dependency* ? not sure

Reply:
Me either, I used the plural since presumably there will be more than 1 new dependency, but even as a native speaker, English is hard.

Reply:
I think this is fine as is.